src/ServiceStack/DependencyInjection/DependencyService.cs

```diff
@@ -37,22 +39,82 @@ public DependencyService()
+                {
+                    registration = registration.FindConstructorsWith(type => type.GetConstructors(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic));
+                }
+                registration = SetRegistrationLifetime(registration, sharing);
+            }
+        }
+
+        public void RegisterAsType(Type implementingType, Type registrationType, Sharing sharing, bool includeNonPublicConstructors)
+        {
+            RegisterAsType(implementingType, new Type[] {registrationType}, sharing, includeNonPublicConstructors);
         }
 
-        public void RegisterTypeAsInterface(Type classType, Type interfaceType, Sharing sharing = Sharing.None)
+        public void RegisterAsType(Type implementingType, Type[] registrationTypes, Sharing sharing, bool includeNonPublicConstructors)
         {
-            var registration = _containerBuilder.RegisterType(classType).As(interfaceType);
-            SetSharing(registration, sharing);
+            var registration = _containerBuilder.RegisterType(implementingType).As(registrationTypes);
+            if (includeNonPublicConstructors)
+            {
+                registration = registration.FindConstructorsWith(type => type.GetConstructors(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic));
+            }
+            registration = SetRegistrationLifetime(registration, sharing);
+        }
+
+        public void RegisterSingletonInstance(object classInstance, bool registerAsImplementedInterfaces)
+        {
+            var classType = classInstance.GetType();
+            var registrationTypes = new Type[] { classType };
+            if (registerAsImplementedInterfaces)
+            {
+                var interfaceTypes = classType.GetInterfaces();
+                registrationTypes = registrationTypes.Union(interfaceTypes).ToArray();
+            }
+            var registration = _containerBuilder.Register(c => classInstance);
+            registration = registration.As(registrationTypes);
+            registration = SetRegistrationLifetime(registration, Sharing.Singleton);
         }
 
-        public void RegisterSingletonInstance(object classInstance, Type classOrInterfaceType)
+        private IRegistrationBuilder<TLimit, TActivatorData, TRegistrationStyle> SetRegistrationLifetime<TLimit, TActivatorData, TRegistrationStyle>(
+            IRegistrationBuilder<TLimit, TActivatorData, TRegistrationStyle> registration,
+            Sharing sharing)
         {
-            var registration = _containerBuilder.Register(c => classInstance).As(classOrInterfaceType);
-            SetSharing(registration, Sharing.Singleton);
+            switch (sharing)
+            {
+                case Sharing.None:
+                    registration = registration.InstancePerDependency();
+                    break;
+                case Sharing.PerRequest:
+                    registration = registration.InstancePerLifetimeScope();
```

Comment:
We create a scope with each HTTP request, which is done when calling the `CreateResolver()` method below.  The `DependencyResolver` object implements `IDisposable` and is cleaned up at the end of the HTTP request by ServiceStack.